packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml

```
@@ -166,19 +170,42 @@ processors:
 
   - foreach:
       field: killChainPhases.edges
+      ignore_missing: true
       processor:
         append:
           field: opencti.indicator.kill_chain_phase
           value: "[{{{_ingest._value.node.kill_chain_name}}}] {{{_ingest._value.node.phase_name}}}"
   - remove:
       field: killChainPhases
 
-  - foreach:
-      field: externalReferences.edges
-      processor:
-        append:
-          field: opencti.indicator.external_reference
-          value: "{{{_ingest._value.node}}}"
+  - script:
+      description: Move external reference nodes to their target location, without nulls
+      lang: painless
+      source: |
+        ArrayList edges = ctx.externalReferences.edges;
+        if (edges == null) {
+          return;
+        }
+        for (int i = 0; i < edges.length; i++) {
+          if (!ctx.opencti?.indicator?.containsKey('external_reference')) {
```

This is not safe; the `?.` operator stops the derefs being unsafe, but then we get to a look up for the method on `null`, and bang!
```suggestion
          if (ctx.opencti?.indicator != null && !ctx.opencti.indicator.containsKey('external_reference')) {
```

---

Here the last null safe operator replaces a plain `.containsKey()` method call. My rule of thumb is that `?.` will make the thing immediately to its right safe - either a field access or a method call.

This script returns "safely got a null":
```
POST _scripts/painless/_execute
{
  "script": {
    "params": {
      "key": null
    },
    "source": """
      if (params.key?.nonKey?.nonMethod("arg")?.nonMethod2("arg") == null) {
        return "safely got a null";
      }
    """
  }
}
```

---

Actually, I see there's a problem. The `...?.containsKey('external_reference')` is okay, but `null` as the result of an `if` condition is not okay. So I fixed that and other occurrences like this
```
if (ctx.opencti?.indicator?.containsKey('external_reference') == true) {
  ...
}
```

---

Thanks, that's very informative; I've learned something new.